tests/test_p1_p2.py

```diff
@@ -0,0 +1,66 @@
+import networkx as nx
+from productions import P1, P2
+
+
+def test_p1_p2_flow_should_create_correct_structure():
+    G = nx.Graph()
+    G.add_node(1, label='El', pos=(1/2, 1/2), layer=0)
+
+    assert True == P1.apply(G)
+    assert True == P2.apply(G)
+    assert True == P2.apply(G)
+    assert True == P2.apply(G)
+
+    expected_nodes = [
+        {'label': 'el', 'pos': (0.5, 0.5), 'layer': 0},
+        {'label': 'i', 'pos': (0.5, 0.5), 'layer': 1},
+        {'label': 'E', 'pos': (0, 0), 'layer': 1},
+        {'label': 'E', 'pos': (1, 0), 'layer': 1},
+        {'label': 'E', 'pos': (0, 1), 'layer': 1},
+        {'label': 'E', 'pos': (1, 1), 'layer': 1},
+        # SECOND LAYER
+        {'label': 'i', 'pos': (0.25, 0.5), 'layer': 2},
+        {'label': 'i', 'pos': (0.75, 0.5), 'layer': 2},
+        {'label': 'E', 'pos': (0, 0), 'layer': 2},
+        {'label': 'E', 'pos': (0.5, 0), 'layer': 2},
+        {'label': 'E', 'pos': (0, 1), 'layer': 2},
+        {'label': 'E', 'pos': (0.5, 1), 'layer': 2},
+        {'label': 'E', 'pos': (1, 0), 'layer': 2},
+        {'label': 'E', 'pos': (1, 1), 'layer': 2},
+        # THIRD LAYER
+        {'label': 'I', 'pos': (0.125, 0.5), 'layer': 3},
+        {'label': 'I', 'pos': (0.375, 0.5), 'layer': 3},
+        {'label': 'E', 'pos': (0, 0), 'layer': 3},
+        {'label': 'E', 'pos': (0.25, 0), 'layer': 3},
+        {'label': 'E', 'pos': (0, 1), 'layer': 3},
+        {'label': 'E', 'pos': (0.25, 1), 'layer': 3},
+        {'label': 'E', 'pos': (0.5, 0), 'layer': 3},
+        {'label': 'E', 'pos': (0.5, 1), 'layer': 3},
+
+        {'label': 'I', 'pos': (0.625, 0.5), 'layer': 3},
+        {'label': 'I', 'pos': (0.875, 0.5), 'layer': 3},
+        {'label': 'E', 'pos': (0.5, 0), 'layer': 3},
+        {'label': 'E', 'pos': (0.75, 0), 'layer': 3},
+        {'label': 'E', 'pos': (0.5, 1), 'layer': 3},
+        {'label': 'E', 'pos': (0.75, 1), 'layer': 3},
+        {'label': 'E', 'pos': (1, 0), 'layer': 3},
+        {'label': 'E', 'pos': (1, 1), 'layer': 3},
+    ]
+
+    assert len(G.nodes()) == len(expected_nodes)
+
+    for _, data in G.nodes(data=True):
+        assert data in expected_nodes
```

Comment:
it doesn't test if G is the same as expected_nodes. It only tests if each data from G is in expected_nodes.
```
G data: [
        {'label': 'E', 'pos': (0, 0), 'layer': 1},
        {'label': 'E', 'pos': (0, 0), 'layer': 1}
]
expected_nodes: [
        {'label': 'E', 'pos': (0, 0), 'layer': 1},
        {'label': 'E', 'pos': (1, 0), 'layer': 1}
]
```
G has two nodes that are the same. In this case your test passes.

Reply:
Good catch, removing matched items should work fine

Reply:
There is simpler solution. You can sort both containers and check their equality.

Reply:
Or just make sets from both of them

Reply:
I'm not sure implementing a function that hashes the dicts so they can be sorted is a "simpler solution". Thoughts?

Reply:
> Or just make sets from both of them

you can't because what if list has two nodes that are the same? Then set would delete duplicates and we don't want that.


> I'm not sure implementing a function that hashes the dicts so they can be sorted is a "simpler solution". Thoughts?

You can do that
![obraz](https://user-images.githubusercontent.com/24357718/207946101-ea4fed1e-1024-4f28-a1a8-d33112e95656.png)

Reply:
Yes, I am aware, but it would require me to implement the key method so it takes into consideration all of the fields of the data dict (basically the hashing function I mentioned) and I feel like we don't get much benefit from doing that.
Before someone asks: we need to hash using each field in the dict, if we don't then the arbitrary order in each list will determine if the tests pass.

Reply:
Irrelevant due to the issue pointed out below by @rozchlastywacz